## Aim:
How to link the rule-based extracted CI_TYPE-GEO pairs with the respective Ci failure impacts

**Idea**\
Test using prompt engineering by passing table of CIGEO pairs to GPT-J model.
Steps:
* Load model and apply it always on one chunk of the document to extract CI failure impacts 
* Use prompt engineering to extract time and location of the CI failure (origin) the CI impacts (impact location)
or 
* Pass dataframe of pairs as input to the model
or
* Use few shot prompting with example answers

**Finally:**
* Compaire all approaches of spatial and temporal linking CI failure impacts

In [ ]:
import os
import sys
import subprocess
import importlib

import re
import time
from glob import glob
from pathlib import Path

from io import StringIO

import numpy as np
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
from jinja2 import Environment, FileSystemLoader
from langchain_docling import DoclingLoader
import langdetect

import spacy
from huggingface_hub import login
import transformers
from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    AutoModelForSeq2SeqLM,
    BitsAndBytesConfig,
)

import torch

from pdfminer.high_level import extract_text
from docling.backend.pypdfium2_backend import PyPdfiumDocumentBackend
from docling.datamodel.base_models import InputFormat
from docling.datamodel.pipeline_options import (
    PdfPipelineOptions,
    AcceleratorOptions,
    AcceleratorDevice,
)
from docling.pipeline.standard_pdf_pipeline import StandardPdfPipeline
from docling.document_converter import DocumentConverter, FormatOption


sys.path.append("../")
from src.settings import settings as s
import src.document_cleaning as dc
import src.translation_model as tm
# from src.hf_model_access import get_weight_dir


torch.manual_seed(42)

#  automatic linebreaks and multi-line cells.
pd.set_option("display.max_colwidth", None)
pd.set_option("display.colheader_justify", "left")


test_mode = True
batch_size_speedyGPU = 8 # 8 = max for nvidiva GPU
# play around to find best size to better utilize the GPU (check via nvtop) 

torch.cuda.empty_cache() 
print(torch.cuda.memory_reserved() / 1e9)


Running on local machine
0.0


In [2]:
# os.environ["HF_HOME"] = s.HF_HOME_DIR
try: 
    login(token=os.getenv("HUGGINGFACE_TOKEN"))   # notebook_login
except:
    login(token=os.environ.get("HUGGINGFACE_TOKEN"))  # former HF_TOKEN




Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


## Set paths and vars

In [3]:
# set wd to project root
os.chdir("/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval")



## set path variables
DOCS_DIR = Path(s.PATH_DATA + "text_sources/")
PARSED_TEXT_DIR = Path(s.PATH_DATA + "parsed_documents/")
LLM_OUTPUTS_DIR = Path(s.PATH_DATA + "llm_outputs/")
NER_PATTERNS_FILEPATH = Path(s.NER_PATTERNS_FILEPATH)

# document cleaning
md_dir = Path(PARSED_TEXT_DIR)
md_dir.mkdir(parents=True, exist_ok=True)

# CI GEO pairs
CI_GEO_FILEPATH = Path("./" + s.PATH_DATA + s.CI_GEO_PAIRS_FILENAME)

## store LLM 1 response and prompt
os.makedirs(s.PATH_LLM_DATA, exist_ok=True)
OUTPUT_LLM1_FILEPATH =  Path(s.PATH_LLM_DATA / s.LLM_DATA_FILENAME)




### Set test mode

In [4]:

if test_mode:
    print("Test mode is ON. Using only a small sample of documents for testing.")

    docs_list_sample = [
        # Path(PARSED_TEXT_DIR, "AEMET 2024 - ESTUDIO SOBRE LA SITUACIÓN DE LLUVIAS INTENSAS_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Karakatsani 2023 - Greece economy briefing The economic impact of the recent devastating floods in Greece_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Lloyd's List 2024 - Port of Valencia reopens after devastating floods_cleaned.md"),
        Path(PARSED_TEXT_DIR, "Containerlift 2024 - Valencia Port Resumes Operations Following Devastating Flooding in Spain - Containerlift.co.uk - Transport_Lifting_Shipping_cleaned.md"), 
            Path(PARSED_TEXT_DIR, "ABC 2024 - Traffic jams and flight delays due to heavy rain and lightning storm in Malaga_cleaned.md"),

        #     Path(PARSED_TEXT_DIR, "Koks 2022 - Brief communication_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "European Investment Bank 2025 - Spain_ EIB lends €50 million to Iberdrola to rebuild and climate-proof flood-hit power infrastructure in Valencia_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Wilson 2024 - Flash floods in Spain sweep away cars, disrupt trains and leave several missing _ AP News_cleaned.md"),     
        #     Path(PARSED_TEXT_DIR, "Wildhagen 2013 - Hochwasser_ Wie die Flut Unternehmen lahmlegt_cleaned.md"),

        # # # not Deidda et al, IPCC, Fekete 2025 as it already contains coarse info about many CI impacts
        # Path(PARSED_TEXT_DIR, "ABC 2024 - Traffic jams and flight delays due to heavy rain and lightning storm in Malaga_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "AFP 2022 - The_Vibes_Valencia Airport in Madrid briefly shut as lightning hits runway _ World _ The Vibes_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Artemis 2015 - PERILS finalises Storm Desmond UK flood loss estimate at £604m_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Brown 2010 - Economy feels chill as UK grinds to a halt _ The Independent_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Diakakis 2020 - A systematic assessment of the effects of extreme flash floods on transportation infrastructure and circulation: The example of the 2017 Mandra flood_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "EFE 2024 - The DANA storm, live_ The death toll rises to 158_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Eurelectric 2006 - Impacts of Severe Storms on Electric Grids_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Euronews 2024 - Spain floods_ Death toll rises to 205 as nation braces for more rain _cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Ferlita 2023 - Incendi in Sicilia, ecco cosa accade_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Fink 2004 - The 2003 European summer heatwaves and drought - synoptic diagnosis and impacts_cleaned.md"),
        # Path(PARSED_TEXT_DIR, "Gilbody Dickerson 2024 - Spain floods_ At least 95 people killed including British man near Malaga _ World News _ Sky News_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Kadir 2014 - The Impact of Natural Disasters on Critical Infrastructures - A Domino Effect-based Study_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Kaur 2025 - Authorities suspect arson in 17 wildfires across Dalmatian coast, Croatia - The Watchers_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Keller 2014 - Mapping Natural Hazard Impacts on Road Infrastructure—The Extreme Precipitation in Baden-Württemberg_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Kettle 2020 - Storm Xaver over Europe in December 2013 Overview of energy impacts and North Sea events_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Koks 2019 - Understanding Business Disruption and Economic Losses Due to Electricity Failures and Flooding_cleaned.md"),
        #     Path(PARSED_TEXT_DIR, "Korzilius 2021 Nach der Flut_cleaned.md"),
        ## not part of valid set:
        # Path(PARSED_TEXT_DIR, "Krausmann 2014 - STREST report on lessons learned from recent catastrophic events_cleaned.md"), # > 1800 entries LLMv3.0 incl. hallucinations
        ]


## Test mode
if test_mode:
    search_path = docs_list_sample
    print("Test mode is ON. Using only a small sample of documents for testing.")
else:
    search_path = glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))


Test mode is ON. Using only a small sample of documents for testing.
Test mode is ON. Using only a small sample of documents for testing.


###  Load and initialize spaCy language model

In [5]:
# # spaCy language model for NER and ENTITY LINKING

try: 
    print("Try loading spaCy language model for local machine ...")
    try:
        nlp = spacy.load(s.SPACY_MODEL)
    except (OSError, ValueError):
        print(f"spaCy language model '{s.SPACY_MODEL}' not found. Downloading ...")
        ## loading transformer language model for NER requires additional package
        if (s.SPACY_MODEL.endswith("_trf")):
            !uv add spacy[transformers]
        !uv run python -m spacy download {s.SPACY_MODEL}
        nlp = spacy.load(s.SPACY_MODEL)

except (OSError, ValueError):
    try: 
        print(f"spaCy language model '{s.SPACY_MODEL}' not found. Downloading ...")
        subprocess.check_call(["uv", "pip", "install", "spacy-transformers"])
        subprocess.check_call(["uv", "run", "python3", "-m", "spacy", "download", s.SPACY_MODEL])
        nlp = spacy.load(s.SPACY_MODEL)
    except ValueError:
        print("Try downloading directly via python3 -m spacy ... (and not via uv run ...)")
        subprocess.check_call(["python3", "-m", "spacy", "download", s.SPACY_MODEL])


# !uv run python -m spacy download en_core_web_trf
# nlp = spacy.load("en_core_web_trf")

Try loading spaCy language model for local machine ...


In [6]:
# Initialise spaCy pipeline 
nlp.add_pipe("merge_entities")
nlp.add_pipe("merge_noun_chunks")


# add CI_TYPE patterns to spacy nlp model pipeline
config = {"spans_key": None, "annotate_ents": True, "overwrite": False}
## see for more info: https://spacy.io/usage/rule-based-matching#entityruler
## NOTE EntityRuler is hidden inside .add_pipe()
try:
    ruler = nlp.add_pipe("span_ruler", config=config)
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)
except ValueError:
    print("SpanRuler already exists in pipeline.")
    ruler = nlp.get_pipe("span_ruler")
    ruler.from_disk(s.NER_PATTERNS_FILEPATH)


In [7]:
## Document cleaning

In [8]:
# Docling pipeline configs

accelerator_options = AcceleratorOptions(
    num_threads=4, device=AcceleratorDevice.AUTO
)  # use GPU + multi-threading
pipeline_options = PdfPipelineOptions()
pipeline_options.do_ocr = True
pipeline_options.do_table_structure = (
    True  # identify tables as such just not to have them in the TextItems later
)
pipeline_options.accelerator_options = accelerator_options
pipeline_options.force_backend_text = True




# setup converter for PDF and markdown
converted = DocumentConverter(
    allowed_formats=[InputFormat.PDF, InputFormat.MD],
    format_options={
        InputFormat.PDF: FormatOption(
            pipeline_cls=StandardPdfPipeline,
            pipeline_options=pipeline_options,
            backend=PyPdfiumDocumentBackend,
        ),
    },
)


In [9]:
print( "Number of documents to process:", len(os.listdir(DOCS_DIR)) )


# convert the different layouts of the pdf files into unified markdown format incl. sub/section titles, tables, caption text etc
for pdf_filename in os.listdir(DOCS_DIR):
    if pdf_filename.endswith(".pdf"):

        md_filename = f"{Path(pdf_filename).stem}.md"

        pdf_filepath = os.path.join(DOCS_DIR, Path(pdf_filename))
        md_filepath = os.path.join(PARSED_TEXT_DIR, Path(md_filename))
        cleaned_md_filepath = md_filepath.replace(".md", "_cleaned.md")

        if os.path.exists(md_filepath):
            print(
                f"Markdown file '{md_filepath}' already exists. Skipping conversion and cleaning."
            )
            continue

        start_time = time.time()
        print(f"\nFetching: {pdf_filename}")

        print("Remove reference section")
        pdf_text = extract_text(pdf_filepath)
        pdf_text_no_refs = dc.remove_references(pdf_text)

        print("Removing URLs") # LangExtract tries to open these URLs when they occur in the document text
        pdf_text_no_refs_urls = re.sub(r"http\S+", "", pdf_text_no_refs) 

        # FIXME remove workaround of saving pdf as markdown and reading it again as Docling.Document
        # FIXME with DocLoader
        # loader = DoclingLoader(md_filepath)
        # md_text = loader.load()
        with open(md_filepath, "w", encoding="utf-8") as f:
            f.write(pdf_text_no_refs_urls)
        print("Converting Markdown to text...")
        md_text = converted.convert(md_filepath)

        print("Removing headers and footers\n")
        md_text_cleaned = dc.remove_headers_footers(md_text)

        print(f"Saving parsed and cleaned document as markdown to: {cleaned_md_filepath}")
        md_text_cleaned.document.save_as_markdown(cleaned_md_filepath)

        end_time = time.time() - start_time
        print(f"Parsing and cleaning done. Time elapsed: {end_time:.2f} seconds.")


# visual check of removed items
# TODO make as document_cleaning function: print removed items with largest number of chars first
# ## NOTE. high number of chars == more potentially actual text body

# text_items_removed = sorted(text_items_to_drop_visualization, key=lambda x: -x[0])
# for i in text_items_removed[:50]:
#     print(i) # -->  also subsection titles were removed partly





Number of documents to process: 49
Markdown file '../data/parsed_documents/UNEP 2013 - Impacts of summer 2003.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Stamataki 2023 - Greece’s record rainfall and flash floods are part of a trend _ PreventionWeb.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Koks 2022 - Brief communication.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/Yamashita 2008 - Analysis  control  and economic impact assessment of major blackout events.md' already exists. Skipping conversion and cleaning.
Markdown file '../data/parsed_documents/AEMET 2024 - ESTUDIO SOBRE LA SITUACIÓN DE LLUVIAS INTENSAS.md' already exists. Skipping co

## LLama

###  Prompt engineering

In [10]:
question_1 = "Which infrastructure failures are mentioned in the text? Categorize the output by the type of infrastructure, the location, and the type of damage."

In [11]:
env = Environment(loader=FileSystemLoader("./prompt_templates/"))
template = env.get_template("ci_loc_direct_impacts.txt")


### Init Llama application function 
* extract CI-GEO pairs (ie. most likely geolocation of each CI_TYPE )
* pass NER table to prompt for evaluating and improving LLM response
* Test approach by applying it on three cleaned documents


In [12]:
# # empty CUDA cache
import gc
import torch

gc.collect()

torch.cuda.empty_cache()
torch.no_grad()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))



In [13]:
# init class for decoder and tokenizer


class DecoderModel:

    def __init__(self, model_name: str ="meta-llama/Llama-2-7b-chat-hf"):
        
        try: 
            login(token=os.getenv("HUGGINGFACE_TOKEN"))   # notebook_login
        except:
            login(token=os.environ.get("HUGGINGFACE_TOKEN"))  # former HF_TOKEN

        base_dir =  s.HF_HOME_DIR   # use default dir in .cache/
        model_dir = base_dir # / f"models--{model_name.replace("/", "--")}"  # is already .._mirror/hub/
        
        print(model_dir)

        # quantization config
        # Load model with 4-bit quantization if applicable (use 4-bit integer instead of 32b floats) --> reduce the required VRAM for model application
        # see, https://huggingface.co/docs/transformers/quantization
        bnb_config = BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type="nf4",
            bnb_4bit_compute_dtype=torch.float16,
        )

        self.pipeline, self.tokenizer = self.initialize_model(
            model_name, model_dir, bnb_config
        )
        

    def initialize_model(self, model_name: str, model_dir: str = None, bnb_config=None):
        
        # Model and Tokenizer initialization
        if not os.path.exists(model_dir):
            print("Model directory not found. Downloading model...")
            os.makedirs(model_dir, exist_ok=True)

            device = transformers.infer_device()
            print(f"Using device: {device}")
            model = AutoModelForCausalLM.from_pretrained(
                model_name,
                dtype="auto",
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                # max_memory={0: "2GB", 1: "10GB"},  # distribute memory across GPUs
            )
            model.save_pretrained(model_dir)
            tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
            tokenizer.save_pretrained(model_dir)

            print("Downloaded model and tokenizer")

        else:
            print(f"Using locally saved model from {model_dir}")

            model = AutoModelForCausalLM.from_pretrained(
                model_name,
                cache_dir=model_dir,
                local_files_only=True,  # tp_plan="auto" # set tensor parallel model (ie. splits model on multiple GPU)
                dtype="auto",
                attn_implementation="flash_attention_2",  # use with 4-bit quantization,
                # --> flash attention enables to use much larger sequence lengths without running into OOM issues
                quantization_config=bnb_config,
                # tp_plan="auto",  # automatically use a tensor parallelism plan based on predefined configuration of the model (i.e. partition model on both GPUs)
            )
            # print("Tensor parallel plan:", model._tp_plan)

            tokenizer = AutoTokenizer.from_pretrained(
                model_name,
                use_fast=True,
                cache_dir=model_dir,  # use fast Rust-based tokenizer, when possible
            )

        # reduce further memory usage
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        model = model.to(device)
        model.use_checkpointing = True

        torch.cuda.empty_cache()
        torch.no_grad()

        # Pipeline setup for question answering
        pipeline = transformers.pipeline(  # load model locally from wsl .cache\
            "text-generation",
            model=model,
            tokenizer=tokenizer,
            max_new_tokens=1024, # high max token otherwise output is truncated
            device_map="auto",
        )
        return pipeline, tokenizer

    def generate_response(
        self, question: str, context: list, # chunk_id: int
    ):
    
        rendered_prompt = template.render(
            context=context,  # includes also df_ci_geo info
            question=question,
        )

        # print(f"Generating response for chunk_id: {chunk_id} ...")

        sequences = self.pipeline(
            rendered_prompt,  # jinja template
            max_new_tokens=1024, # use default to not truncate the LLM response
            do_sample=True,
            num_beams=1,  # select token based on probability distribution over entire model’s vocabulary
            # top_k=10,
            # top_p=0.5,
            temperature=0.2,
            batch_size=batch_size_speedyGPU,
            # num_return_sequences=1,
            eos_token_id=self.tokenizer.eos_token_id,
            return_full_text=False,  # allow bullet point answers
        )
        # Extracting and returning the generated text
        return sequences

### Apply Llama on chunks

In [14]:
## Settings
model_name = "meta-llama/Llama-2-7b-chat-hf"
# TODO NOTE test for HPC: "meta-llama/Llama-3.1-8B-Instruct"

time0 = time.time()

## init LLM pipeline
decoder_model = DecoderModel(model_name=model_name)

## init output dataframe
df_responses_all = pd.DataFrame()
sentence_root_list = [] # for visualization of sentence roots and their dependencies
responses_error_list = []


if test_mode:
    search_path = docs_list_sample
    print("Test mode is ON. Using only a small sample of documents for testing.")
else:
    search_path = glob(str(Path(PARSED_TEXT_DIR, "*cleaned.md")))


## Start CI impact extraction
for i, filename in enumerate(search_path):

    src_language_nonengl = None 

    no_documents = len(search_path)
    filepath = Path(filename)
    filename_stem = filepath.stem



    ## load doc
    loader = DoclingLoader(filepath)  # use chunks from Docling.Loader
    doc = loader.load()


    print(f"\n\n ######## -------- Processing document [{i+1}/{no_documents}]: {filepath.name} -------- ######## \n")

    ## extract authors, publication year and title 
    author, year, title = dc.extract_citation_info(filename_stem)
    citation = f"{author} {year}".replace("  ", " ").strip()
    title = title.replace(" - ", "").replace("_cleaned", "").strip()


    # check language of document based on document title
    # check title rather than chunk_text as later might contain also footer or header info in another languages
    
    src_language_doc = langdetect.detect(str(title).lower())  # lower case improves language detection
    
    if src_language_doc != "en":
        supported_languages = ["fr", "de", "es", "it", "itc", "nl"]
        if src_language_doc not in supported_languages:
            # raise ValueError(f"Unsupported source language: {src_language_doc}")
            print(f"Unsupported source language: {src_language_doc}. Continue with extraction on original text")
            continue 
        
        print(f"\n ######## -------- Translating: {src_language_doc} --> en -------- ######## \n")

        # clean up before applying translator
        gc.collect()
        torch.cuda.empty_cache()  # mainyl after training needed, small effect when LLM applied only for infernece
        torch.no_grad()
        
        # return translated Docling Document
        doc = tm.translate_2_english(src_language_doc, doc)



    print(f"\n ######## -------- Relationship extraction: CI-GEO pairs -------- ######## \n")

    ## Create New entity for transport infrastructure and apply it on any doc

    ## TODO make as pydantic class with fixed attributes
    df_ci_geo = pd.DataFrame(
        columns=[
            "citation_id",
            "chunk_id",
            "ci_entity",
            "geo_entity",
            "case_type",
            "chunk_text",
        ]
    )

    for chunk_no, chunk in enumerate(doc):


        ## preprocess  TODO move to document cleaning workflow + dc.funcs
        chunk.page_content = chunk.page_content.replace("\n", " ")
        chunk.page_content = chunk.page_content.replace("- ", "-") # TODO test if ("- ", "") is better


        ## CI-GEO extraction
        ## apply customized NLP pipeline -> returns same chunk.text but with CI_entities annotations and without linebreak symbols
        for nlp_chunk in nlp.pipe([chunk.page_content]):

            ci_entities = [ent for ent in nlp_chunk.ents if ent.label_ in ["CI_TYPE"]]

            # check if chunk has no CI_TYPE entities
            if not ci_entities:
                # print(f"Chunk [{chunk_no}], CI_TYPE entities found: 0")
                # print(f"\nChunk [{chunk_no}], No. CI_TYPE and FAC entities: {len(ci_entities)}")
                continue
      

            print(f"Chunk [{chunk_no}], CI_TYPE entities found: {len(ci_entities)}")
            

            for ci_ent in ci_entities:
                # print("Ci_ent:", ci_ent.text)

                # Get the root of the sentence 
                sentence_root = ci_ent.sent.root

                if sentence_root.dep_ in ("ROOT", "dobj", "conj"):  # TODO probably enough when set only to ROOT 
                    # sentence_root_list.append(sentence_root.dep_) # for visualization of sentence roots and their dependencies
                    head = ci_ent.root.head
                    # print(f" `{sentence_root, sentence_root.dep_}` is sentence_root, \n `{head}` the direct head of ci entity `{ci_ent}`")

                    subjs = list(sentence_root.lefts) + list(sentence_root.rights)
                    for subj in subjs:
                        for descendant in subj.subtree:

                            # check if descendant is a location entity
                            if descendant.ent_type_ in ("GPE", "LOC"):
                                descendant_loc = descendant
                                print("  ", ci_ent.text, "->", descendant_loc.text)
                                # assert subj is descendant_loc or subj.is_ancestor(descendant_loc)

                                ## Special case handling - one root but multiple ci-geo pairs
                                try:
                                    first_ci_ancestor = [ancestor.text for ancestor in descendant_loc.ancestors if ancestor.ent_type_ in ("CI_TYPE", "FAC")][0]
                                    if ci_ent.text == first_ci_ancestor:
                                        print(
                                            "descendant of location:", descendant_loc.text, descendant_loc.dep_, 
                                            "\nits ancestors:", [ancestor.text for ancestor in descendant_loc.ancestors] # all parents up to root.head
                                        )
                                        print("  ", ci_ent.text, "-->", descendant_loc.text)
                                        
                                        ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                                        result_dict = {
                                            "citation_id": citation,
                                            "chunk_id": chunk_no,
                                            "ci_entity": ci_ent.text,
                                            "geo_entity": descendant_loc.text,
                                            "case_type": "special",
                                            "chunk_text": nlp_chunk.text,
                                        }
                                        df_ci_geo = pd.concat(
                                            [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                                        )

                                ## usual case handling
                                except IndexError:
                                    ## write as dict entry incl chunk_id, ci_entity, geo_entity, distance
                                    result_dict = {
                                        "citation_id": citation,
                                        "chunk_id": chunk_no,
                                        "ci_entity": ci_ent.text,
                                        "geo_entity": descendant_loc.text,
                                        "case_type": "usual",
                                        "chunk_text": nlp_chunk.text,
                                    }
                                    df_ci_geo = pd.concat(
                                        [df_ci_geo, pd.DataFrame([result_dict])], ignore_index=True
                                    )

                else:
                    # print("\nNo CI_TYPE or FAC entities found in this chunk.")
                    continue

    ## post-process of DF CI-GEO pairs
    df_ci_geo = df_ci_geo.drop_duplicates(
        subset=["citation_id", "chunk_id", "ci_entity", "geo_entity","case_type", "chunk_text"]
        )# .reset_index(drop=True, inplace=True)


    print(f"\n  #############  -------- Text-2-Data: {filepath.name} -------- #############  \n")

    df_responses = pd.DataFrame(
        columns=[
            "citation_id",
            "chunk_id",
            "infrastructure_type",
            "damage",
            "location",
            "ci_entity",
            "geo_entity",
            "case_type",
            "chunk_text"
        ]
    )

    ## apply decoder on each chunk in document
    ## TODO replace iteration by loading entire document and use recursive chunking from langchain
    for j, chunk in enumerate(doc):

        if df_ci_geo.loc[df_ci_geo["chunk_id"] == j].empty:
            context = [
                {
                    "text": chunk.page_content,
                    "citation": citation,
                    "title": filename_stem, 
                    "ci_locations": None,
                },
            ]
        else:  # TODO disolve if else clause by making it in ci_locations: if df_ci_geo.chunk=j, xx, else None
            context = [
                        {
                            "text": chunk.page_content,
                            "citation": citation,
                            "title": filename_stem, 
                            # TODO check if chunk id is unique per doc and if multiple ci_loc pairs eixist (then just tke first one)
                            "ci_locations": df_ci_geo.loc[df_ci_geo["chunk_id"] == j],
                        },
                    ]

        # apply LLM
        response = decoder_model.generate_response(
            question=question_1, context=context, # chunk_id=j
        )

        ## postprocess response
        resp = response[0]["generated_text"].replace("\n", "")
        try:
            # fix missing bracket at beginning and end of response by splitting at first/last complete entry
            if "[{" not in resp:
                resp = "[{" + resp.split("{", 1)[1]
            if "}]" or "},]" not in resp:
                resp = resp.rpartition('}')[-3] + "}]"

            # elif "]" not in resp:
            #     resp = (resp.split("]")[0] + "]")
            # elif "[" not in resp:
            #     resp = ("[" + resp.split("[")[1]) 

            # remove potential text outside - no matter if it exists or not
            resp = ("[" + resp.split("[")[1]) 
            resp = (resp.split("]", 1)[0] + "]") 

            # save LLM response for each chunk  in dataframe for each document
            df_resp = pd.read_json(StringIO(resp))
            df_resp["citation_id"] = context[0]["citation"]  # add citation info
            df_resp["chunk_id"] = j  # add chunk id as identifier
            df_resp["ci_entity"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "ci_entity"] 
            df_resp["geo_entity"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "geo_entity"]
            df_resp["case_type"] = df_ci_geo.loc[df_ci_geo["chunk_id"] == j, "case_type"]
            df_resp["chunk_text"] = context[0]["text"]  # add (translated) chunk text for tracing back LLM response

            df_responses = pd.concat([df_responses, df_resp], ignore_index=True)
        
        except (IndexError, ValueError) as e:
            print(f"Cannot add response: {e}, \n   Faulty response (before postprocessing): {response[0]['generated_text'].replace('\n', '')}")
            # faullty response: e.g.  .., "location": "V" on satellite and online on radar"}, { ...}, {}
            responses_error_list.append({
                "citation_id": citation,
                "chunk_id": j,
                "response": response[0]['generated_text'].replace('\n', ''),
                "error": str(e)
            })


    df_responses_all = pd.concat([df_responses_all, df_responses], ignore_index=True)


    # clean up after each document
    gc.collect()
    torch.cuda.empty_cache()  # mainly needed after training, small effect when LLM applied only for inference
    torch.no_grad()

print(f"\n\n ######## -------- CI impact extraction took {(time.time() - time0) / 60} minutes -------- ######## \n\n")

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub
Using locally saved model from /home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Device set to use cuda:0


Test mode is ON. Using only a small sample of documents for testing.


 ######## -------- Processing document [1/5]: Karakatsani 2023 - Greece economy briefing The economic impact of the recent devastating floods in Greece_cleaned.md -------- ######## 


 ######## -------- Relationship extraction: CI-GEO pairs -------- ######## 

Chunk [0], CI_TYPE entities found: 1
Chunk [3], CI_TYPE entities found: 3
Chunk [6], CI_TYPE entities found: 1
   the roadshow -> Athens
descendant of location: Athens pobj 
its ancestors: ['in', 'organized', 'the roadshow', ' ', 'investors', ' ', 'consider']
   the roadshow --> Athens

  #############  -------- Text-2-Data: Karakatsani 2023 - Greece economy briefing The economic impact of the recent devastating floods in Greece_cleaned.md -------- #############  



Token indices sequence length is longer than the specified maximum sequence length for this model (585 > 512). Running this sequence through the model will result in indexing errors




 ######## -------- Processing document [2/5]: Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung_cleaned.md -------- ######## 


 ######## -------- Translating: de --> en -------- ######## 

/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub
Using device: cuda
Using locally saved model from /home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/notebooks/huggingface_mirror/hub


/home/a-buch/Documents/TUB_DWN/_PROJECTS/CI-impacts-information-retrieval/.venv/lib/python3.12/site-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


Detected language for chunk 0: de
 Source text:  Juni-Hochwasser 2013 in Mitteleuropa -Fokus Deutschland Bericht 2: Auswirkungen und Bewältigung1  (Der Bericht berücksichtigt die Entwicklung bis zum 11. Juni 2013, 16:00 Uhr)  Autoren: Bijan Khazai, Tina Bessel, Stella Möhrle, André Dittrich, Kai Schröter, Bernhard Mühr, Florian Elmer, Tina Kunz-Plapp, Werner Trieselmann, Michael Kunz  Deichbruch bei Fischbeck (Foto, 10.Juni 2013) Landkreis Stendal, im Norden Sachsen-Anhalts (Foto: Berns-Volker Brahms / Volksstimme)2  1 Der Bericht wurde am 13.06.2013 publiziert. Abbildung 1, Seite 6; Tabelle 13 und Abbildung 12, Seite 20 wurden am 19.06.2013 überarbeitet. 
 Translated text:  June 2013 in Central Europe -Focus Germany Report 2: Impacts and coping1 (The report takes into account the development until June 11, 2013, 4:00 p.m.) Authors: Bijan Khazai, Tina Bessel, Stella Möhrle, André Dittrich, Kai Schröter, Bernhard Mühr, Florian Elmer, Tina Kunz-Plapp, Werner Trieselmann, Michael Kunz Dei

You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset



  #############  -------- Text-2-Data: Khazai 2013 - Juni-Hochwasser 2013 in Mitteleuropa - Fokus Deutschland Bericht 2 Auswirkungen und Bewältigung_cleaned.md -------- #############  



OutOfMemoryError: CUDA out of memory. Tried to allocate 20.00 MiB. GPU 0 has a total capacity of 5.67 GiB of which 3.88 MiB is free. Process 5344 has 112.18 MiB memory in use. Including non-PyTorch memory, this process has 5.52 GiB memory in use. Of the allocated memory 4.97 GiB is allocated by PyTorch, and 420.93 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

In [ ]:
print("Chunk with erroneous responses:", responses_error_list.__len__())
df_responses_error = pd.DataFrame(responses_error_list)
df_responses_error#.tail(3)


Chunk with erroneous responses: 0


""


In [ ]:
responses_error_list #df_responses_all

[]

### Saving

In [ ]:
# PATH_LLM_DATA: Path = Path(s.PATH_DATA + "llm_outputs/")
# LLM_DATA_FILENAME: str = "llm_1_robustified.csv"

# OUTPUT_LLM1_FILEPATH =  Path(PATH_LLM_DATA / LLM_DATA_FILENAME)#.replace(".csv", "_v2.csv"))


In [ ]:
safety_df = df_responses_all.copy()



# save LLM 1 output to disk along with prompt text
if not os.path.isfile(OUTPUT_LLM1_FILEPATH):

    print(f"Saving prompt, LLM response and erroneous responses [.txt, .csv] to {OUTPUT_LLM1_FILEPATH} ...")

    with open(f"prompt_{OUTPUT_LLM1_FILEPATH.stem}.txt", "w") as f:
        f.write(template.render(context=context, question=question_1))
    df_responses_all.to_csv(OUTPUT_LLM1_FILEPATH, index=False)
    df_responses_error.to_csv(Path(OUTPUT_LLM1_FILEPATH.parent, f"errors_{OUTPUT_LLM1_FILEPATH.stem}.csv"), index=False)

elif os.path.isfile(OUTPUT_LLM1_FILEPATH) and not os.path.isfile(Path(OUTPUT_LLM1_FILEPATH.parent, f"{OUTPUT_LLM1_FILEPATH.stem}_v2.csv")):
    # If the original files exists but the v2 file doesn't, create the v2 file
    with open(f"prompt_{OUTPUT_LLM1_FILEPATH.stem}_v2.txt", "w") as f:
        f.write(template.render(context=context, question=question_1))
    df_responses_all.to_csv(Path(OUTPUT_LLM1_FILEPATH.parent, f"{OUTPUT_LLM1_FILEPATH.stem}_v2.csv"), index=False)
    df_responses_error.to_csv(Path(OUTPUT_LLM1_FILEPATH.parent, f"errors_{OUTPUT_LLM1_FILEPATH.stem}_v2.csv"), index=False)
else:
    print(f"Output file {Path(OUTPUT_LLM1_FILEPATH).stem} already exists. Skip saving to avoid overwriting ...")






Saving prompt, LLM response and erroneous responses [.txt, .csv] to ../data/llm_outputs/llm_1_updprompt.csv ...


Save as pyarrow incl dtypes as pyarraws-. Saving also dtypes a spyarrow objectes leads to quicker loading and processing

In [ ]:
# pd.options.future.infer_string = True

df_responses_all_pyarrow = df_responses_all.astype( dtype="string[pyarrow]")
df_responses_all_pyarrow = pa.Table.from_pandas(df_responses_all_pyarrow)
pq.write_table(df_responses_all_pyarrow, OUTPUT_LLM1_FILEPATH.with_suffix(".parquet"))  



## Evaluation - only manually

In [30]:
df_responses_all.sort_values(by=["ci_entity"])[:50]  # == "Spain’s second-biggest port"]

,citation_id,chunk_id,infrastructure_type,damage,location,ci_entity,geo_entity,case_type,chunk_text
4443,Korzilius 2021,0,hospitals,severely damaged,North Rhine,68 hospitals,North Rhine,usual,"In the middle of July, continuous heavy rains caused small rivers to turn into torrential streams in North Rhine-Westphalia and Rhineland-Palatinate. Meter-high flood waves devastated villages, cities and entire land stings. In North Rhine alone, 145 doctor's practices and 68 hospitals across the country are affected differently by floods. According to initial estimates by the NRW Ministry of Health, the damage is estimated to be more than 100 million euros."
211,AEMET 2024,49,bridges,heavily destroyed,Balearic Islands,AEMET stations,the Balearic Islands,special,"84.4 Table 4.2. Most remarkable precipitations during October 28, 2024 at AEMET stations in the Balearic Islands. Accumulated in mm in one, six, twelve and twenty-four hours. -The most remarkable warnings issued for the areas with the greatest impact in the Balearic community were the following:"
227,AEMET 2024,52,power_lines,heavily destroyed,Castellón,AEMET stations,Castellón,special,"Catellón province CASTELLON Table 4.3. Most remarkable precipitations during October 28, 2024 at AEMET stations in the province of Castellón. Accumulated in mm in one, six, twelve and twenty-four hours. -The warnings issued for the affected predictive areas were the following:"
329,AEMET 2024,75,Road,outages,Malaga,AEMET stations,Malaga,special,"There were persistent precipitations that caused very important accumulations in the region of Sol and Guadalhorce. The Guadalhorce river overflowed to the height of Cartama, dragging cars and flooding farms. A British citizen died and there were hundreds of incidents and rescues. The episode came accompanied by an exceptional electrical activity. The stations of Álora (Málaga) and the meteorological center in the city of Malaga recorded in twelve accumulated hours of 167 mm and 84 mm, respectively. Table 4.7. Most notable precipitations during the 29 October 2024 at AEMET stations in the province of Malaga. Accumulated in mm in one, six, twelve and twenty-four hours."
378,AEMET 2024,80,Dams,heavily damaged,Murcia,AEMET stations,Caravaca,usual,"The precipitations collected during the day exceeded 100 mm in an ellipse with a major axis extending from Caravaca de la Cruz (Murcia), to Molinicos (Albacete), in the Sierra del Segura and Alcaraz, exceeding the 200 mm north-west of Caravaca. At AEMET stations, up to 152.6 mm were accumulated in Caravaca, and 149.6 mm in the station of the Embalse de la Fuensanta (Albacete). At the stations of SUREMET 29 quantities of the order of 250 mm were accumulated in Sierra del Fronton and Rincón de los Huertos, both in Moratalla (Murcia)."
4889,Krausmann 2014,60,Ground slabs,settlement,Gölcük,Derince port,Gölcük,usual,Fig. 6.25 Collapse of ground slabs due to settlement at the Navy port in Gölcük (left) and damage at Derince port (right)........................................................................... 86 \nFig. 6.26 Damage at the navy port in Gölcük (left) and failed column at SEKA port (right) .. 86
912,Khazai 2013,26,evacuation,disrupted,Baden-Württemberg,Numerous dike breaks,the Elbe,special,"The impact of this flood on people, traffic and the economy is great. Numerous dike breaks, for example in Bavaria and along the Elbe, resulted in large-scale regional flooding. In many places, their homes, houses and places had to leave due to evacuation measures. In the Salzlandkreis in Saxony-Anhalt, a voluntary helper died while filling the sandbags and a woman during an evacuation. In Baden-Württemberg, three people, including a firefighter, have been killed since the start of the flood. Up to now, at least 24 victims and 5 missing persons have been killed in all affected flood areas in the Czech Republic, Germany, Austria and Switzerland.3"
911,Khazai 2013,26,sandbags,disrupted,Salzlandkreis,Numerous dike breaks,Ba

### Manual comparison CI_location_table vs LLm response


#### chunk 5
In Germany, road and railway infrastructure was severely damaged as documented exemplarily in Fig. 1. Cost estimates reach up to EURO 2 billion Euro (MDR, 2021). More than 130 km of motorways were closed directly after the event, of which 50 km were still closed two months later, with an estimated repair cost of EUR 100 million (Hauser, 2021). Of the 112 bridges in the flooded 40 km of the Ahr valley (Rhineland-Palatinate), 62 bridges were destroyed, 13 were severely damaged and only 35 were in operation a month after the flood event (MDR, 2021). Over 74 km of roads, paths and bridges in the Ahr valley have been (critically) damaged. In some cases, repairs are expected to take months to years (Zeit Online, 2021). For example, major freeway sections, including parts of the A1 motorway, were closed until early 2022 (24Rhein, 2022). In addition, about 50 000 cars were damaged, causing insurance claims of some EUR 450 million (ADAC, 2021). The German railway provider Deutsche Bahn expects asset damages of around EUR 1.3 billion. Among other things, 180 level crossings, almost 40 signal'

In [ ]:
doc[5].page_content

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"] == 5]

In [ ]:
df_responses[df_responses["chunk_id"] == 5]

In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

### chunk 13 +14
'We found no information regarding direct impact on solid-waste facilities as a result of the flood event. However, there is a large pressure on the solid-waste sector to clean the affected areas; 1 month after the event, we observed dozens of large temporary waste fills and frequent incidences of oil pollution in Rhineland-Palatinate during a field visit. In the Ahrweiler district alone, the flood caused as much solid waste as normally would be collected over 30 years. In Belgium, the amount of solid waste is estimated around 160 000 t, stored at several places, such as the abandoned highway track A601. This highway has been used for approximately 9 months as a temporary storage for debris (Couplez, 2022). In the Netherlands, there have been primarily problems with waste deposits along the river banks, which is mostly the solid waste transported by the river from further upstream. Thousands of tonnes of tree debris (logs and\nNat. Hazards Earth Syst. Sci., 22, 3831–3838, 2022\nE. E. Koks et al.: Flood impacts to infrastructure'


'of running water and electricity (Ärzte Zeitung, 2021). After 1.5 months, medical care was guaranteed again in the most affected regions in Rhineland-Palatinate (Hochwasser Ahr, 2021c). In the state of North Rhine-Westphalia, approximately 68 hospitals have been affected, of which several have been affected severely and will take at least 1.5 years to be rebuilt (Fig. 2). Direct damages are estimated to be at least EUR 100 million to repair all medical facilities (Korzilius, 2021). In the town of Eschweiler (Germany), for example, the basement of the hospital was flooded, as well as the outbuildings and the entire outdoor area. The power supply collapsed, the entire building technology was destroyed and some 300 patients had to be evacuated by helicopter. Property damage is expected to be around EUR 50 million. Within 3.5 weeks, the hospital was partly operational, and within 3 months, all hospital operations continued normally (SAH Eschweiler, 2021). The Mutterhaus Ehrang hospital in Trier (Germany) is now permanently closed as the hospital is too severely damaged to rebuild. Furthermore, in the region of Rhineland-Palatinate (Germany), 19'


In [ ]:
doc[14].page_content

In [ ]:
df_ci_geo[df_ci_geo["chunk_id"].isin([13, 14])]

In [ ]:
df_responses[df_responses["chunk_id"].isin([13, 14])]

In [ ]:
print(
    f"CI_TYPE \n LLM responses:\n {df_responses.infrastructure_type.unique()}, \n\n NER pairs:\n {df_ci_geo.ci_entity.unique()}"
)

In [ ]:
print(
    f"LOCATION:\n LLm responses:\n {df_responses.location.unique()}, \n\n NER pairs:\n {df_ci_geo.geo_entity.unique()}"
)

## Test alternative approaches for entity linking / relation extraction

In [ ]:
import torch
import gc

print(torch.cuda.memory_summary(device=None, abbreviated=False))
# # empyty CUDA cache
gc.collect()

torch.cuda.empty_cache()
torch.no_grad()
# print(torch.cuda.memory_summary(device=None, abbreviated=False))